File: lib/filter/first_order_filter.dart
import '../native_cpp.dart';
import 'dart:ffi'; // For FFI

/// T curOutput = curInput * alpha + prevOutput_ * (1.0 - alpha);
final double Function(double curInput) firstOrderFilterCalc = cppRoboticsLib
    .lookup<NativeFunction<Double Function(Double)>>('FirstOrderFilterCalc')
    .asFunction();

/// params[0] = config.cutOffFreqHz;
///
/// params[1] = config.timeConstantSec;
final Pointer<Double> Function() firstOrderFilterGetParams = cppRoboticsLib
    .lookup<NativeFunction<Pointer<Double> Function()>>(
        'FirstOrderFilterGetParams')
    .asFunction();

final void Function(double cutOffFreqHz, double timeConstantSec)
    firstOrderFilterSetParams = cppRoboticsLib
        .lookup<
            NativeFunction<
                Void Function(Double cutOffFreqHz,
                    Double timeConstantSec)>>('FirstOrderFilterSetParams')
        .asFunction();

final void Function() firstOrderFilterResetFilter = cppRoboticsLib
    .lookup<NativeFunction<Void Function()>>('FirstOrderFilterResetFilter')
    .asFunction();

final void Function() firstOrderFilterResetPrevValues = cppRoboticsLib
    .lookup<NativeFunction<Void Function()>>('FirstOrderFilterResetPrevValues')
    .asFunction();

File: cppRobotics/extern/filter/FirstOrderFilter.cpp
#include "FirstOrderFilter.h"

static FirstOrderFilter<double> firstOrderFilter(50);
static double params[2];

double FirstOrderFilterCalc(double curInput) {
  return firstOrderFilter.Calc(curInput);
}

double* FirstOrderFilterGetParams() {
  FirstOrderFilterConfig config = firstOrderFilter.GetParams();
  params[0] = config.cutOffFreqHz;
  params[1] = config.timeConstantSec;
  return params;
}

void FirstOrderFilterSetParams(double cutOffFreqHz, double timeConstantSec) {
  FirstOrderFilterConfig config = {cutOffFreqHz, timeConstantSec};
  firstOrderFilter.SetParams(config);
}

void FirstOrderFilterResetFilter() {
  firstOrderFilter.ResetFilter();
}

void FirstOrderFilterResetPrevValues() {
  firstOrderFilter.ResetPrevValues();
}
